cmake/OpenCVDetectCUDA.cmake

```diff
@@ -71,19 +72,22 @@ if(CUDA_FOUND)
       endif()
       unset(_header_result CACHE)
     endmacro()
-    ocv_cuda_SEARCH_NVCUVID_HEADER("nvcuvid.h" HAVE_NVCUVID_HEADER)
-    ocv_cuda_SEARCH_NVCUVID_HEADER("dynlink_nvcuvid.h" HAVE_DYNLINK_NVCUVID_HEADER)
-    find_cuda_helper_libs(nvcuvid)
-    if(WIN32)
-      find_cuda_helper_libs(nvcuvenc)
-    endif()
-    if(CUDA_nvcuvid_LIBRARY AND (${HAVE_NVCUVID_HEADER} OR ${HAVE_DYNLINK_NVCUVID_HEADER}))
-      # make sure to have both header and library before enabling
-      set(HAVE_NVCUVID 1)
-    endif()
-    if(CUDA_nvcuvenc_LIBRARY)
-      set(HAVE_NVCUVENC 1)
-    endif()
+	if(WITH_NVCUVID)
+		ocv_cuda_SEARCH_NVCUVID_HEADER("nvcuvid.h" HAVE_NVCUVID_HEADER)
+		ocv_cuda_SEARCH_NVCUVID_HEADER("dynlink_nvcuvid.h" HAVE_DYNLINK_NVCUVID_HEADER)
+		find_cuda_helper_libs(nvcuvid)
+		if(CUDA_nvcuvid_LIBRARY AND (${HAVE_NVCUVID_HEADER} OR ${HAVE_DYNLINK_NVCUVID_HEADER}))
+		  # make sure to have both header and library before enabling
+		  set(HAVE_NVCUVID 1)
+		endif()
+	endif()
+	if(WITH_NVCUVENC)
+		ocv_cuda_SEARCH_NVCUVID_HEADER("nvEncodeAPI.h" HAVE_NVCUVENC_HEADER)
+		find_cuda_helper_libs(nvencodeapi)
```

Comment:
It does not work for Linux. Linux version is named differently: https://docs.nvidia.com/video-technologies/video-codec-sdk/release-notes/